packages/next/client/route-announcer.tsx

```diff
@@ -0,0 +1,67 @@
+import React, { useEffect, useState, useRef } from 'react'
+import { useRouter } from './router'
+
+export function RouteAnnouncer() {
+  const { asPath } = useRouter()
+  const [routeAnnouncement, setRouteAnnouncement] = useState('')
+
+  // Only announce the path change, but not for the first load because screen reader will do that automatically.
+  const initialPathLoaded = useRef(false)
+
+  // Every time the path changes, announce the route change. The announcement will be prioritized by h1, then title
+  // (from metadata), and finally if those don't exist, then the pathName that is in the URL. This methodology is
+  // inspired by Marcy Sutton's accessible client routing user testing. More information can be found here:
+  // https://www.gatsbyjs.com/blog/2019-07-11-user-testing-accessible-client-routing/
+  useEffect(
+    () => {
+      if (!initialPathLoaded.current) {
+        initialPathLoaded.current = true
+        return
+      }
+
+      let newRouteAnnouncement
+      const pageHeader = document.querySelector('h1')
+
+      if (pageHeader) {
+        newRouteAnnouncement = pageHeader.innerText || pageHeader.textContent
+      }
+      if (!newRouteAnnouncement) {
+        if (document.title) {
+          newRouteAnnouncement = document.title
```

Comment:
Looking back at it, I think the logic is backwards here. We should use the `document.title` first, since this is what would happen a non-SPA scenario upon page change. And if we do not have a page title, we should use the `h1` (if any). Right now it will announce the content of the h1, despite having a perfectly clean `document.title`, which I think is not a good experience.

+ @kyleboss for opinions. :)

Reply:
Taking the liberty to ping @shuding and @olpeh as well for comments. ✨